src/Chainweb/Pact/Backend/ModuleCache.hs

```diff
@@ -0,0 +1,184 @@
+-- cache.
+--
+instance Ord CacheEntry where
+  a `compare` b = cmp _ceTxId (cmp _ceAddy EQ)
+    where
+      cmp acc next = case acc a `compare` acc b of
+        EQ -> next
+        c -> c
+
+-- -------------------------------------------------------------------------- --
+-- ModuleCache
+
+data ModuleCache = ModuleCache
+  { _mcStore :: !(HM.HashMap CacheAddress CacheEntry)
+  , _mcSize :: !Int64
+  , _mcLimit :: !Int64
+  }
+  deriving (Show)
+
+makeLenses 'ModuleCache
+
+emptyCache :: Int64 -> ModuleCache
+emptyCache limit = ModuleCache
+    { _mcStore = mempty
+    , _mcSize = 0
+    , _mcLimit = limit
+    }
+
+isEmpty :: ModuleCache -> Bool
+isEmpty = HM.null . _mcStore
+
+-- | The totall size of all modules in the cache.
+--
+-- Complexity: \(O(1)\)
+--
+size :: ModuleCache -> Int
+size = fromIntegral . _mcSize
+
+-- | Number of modules in the cache.
+--
+-- Complexity: \(O(n)\), linear in the number of entries
+--
+count :: ModuleCache -> Int
+count = HM.size . _mcStore
+
+-- | Decode Module from row data using the module cache.
+--
+checkModuleCache
+    :: Utf8
+        -- ^ Row Key for the module
+    -> ByteString
+        -- ^ row data that contains the encoded module
+    -> TxId
+        -- ^ TxId. This is used as priority for cache eviction. Smaller values
+        -- are evicted first.
+    -> ModuleCache
+    -> (Maybe PersistModuleData, ModuleCache)
+checkModuleCache key rowdata txid = runState $ do
+    readModuleCache txid addy >>= \case
+
+        -- Cache hit
+        Just !x -> return $ Just x
+
+        -- Cache miss: decode module and insert into cache
+        Nothing -> case decodeStrict' rowdata of
+            Nothing -> return Nothing
+            Just !m -> do
+                writeModuleCache txid addy m
+                return $ Just m
+  where
+    addy = mkAddress key rowdata
+
+-- -------------------------------------------------------------------------- --
+-- Internal
+
+mkAddress :: Utf8 -> ByteString -> CacheAddress
+mkAddress (Utf8 key) rowdata = CacheAddress $
+  key <> ":" <> BA.convert (C.hash @_ @SHA512t_256 rowdata)
+
+readModuleCache
+    :: TxId
+        -- ^ TxId of
+    -> CacheAddress
+    -> State ModuleCache (Maybe PersistModuleData)
+readModuleCache txid ca = do
+  mc <- use mcStore
+  forM (HM.lookup ca mc) $ \e -> do
+      mcStore . ix ca . ceTxId .= txid
+      return $! _ceData e
+
+-- | Add item to module cache
+--
+writeModuleCache
+    :: TxId
+    -> CacheAddress
+    -> PersistModuleData
+    -> State ModuleCache ()
+writeModuleCache txid ca v = do
+  let sz = sizeOf SizeOfV1 $ void (_mdModule v)
+  mcStore %= HM.insert ca (CacheEntry txid ca sz v)
+  mcSize %= (+ sz)
+  maintainCache
+
+-- | Prune Cache to meet size limit
+--
+-- Complexity: \(O(n \log(n))\) in the number of entries
+--
+maintainCache :: State ModuleCache ()
+maintainCache = do
+    sz <- use mcSize
+    lim <- use mcLimit
+    e <- gets isEmpty
+    unless (e || sz < lim) $ do
```

Comment:
For `lim > 0`, `e` implies `sz < lim`. Therefor the check of emptiness seem redundant here.

Reply:
agree, remove